File: experiments/go/internal/fidelity/fidelity.go
package fidelity

import (
	"math"
	"os"
	"regexp"
	"sort"
	"strconv"
	"strings"

	"github.com/plinde/skillscope/internal/models"
	"github.com/plinde/skillscope/internal/parser"
)

var (
	wordRe           = regexp.MustCompile(`[a-z0-9][a-z0-9_-]*`)
	triggerSectionRe = regexp.MustCompile(`(?i)triggers?\s+on\s*:`)
	quotedPhraseRe   = regexp.MustCompile(`"([^"]{2,60})"`)
)

// stopwords: common English function words plus generic/domain-neutral
// words called out in the spec. Copied from fidelity.py's _STOPWORDS.
var stopwords = buildStopwords()

func buildStopwords() map[string]struct{} {
	words := []string{
		"a", "an", "the", "and", "or", "but", "if", "then", "else", "when",
		"while", "for", "to", "of", "in", "on", "at", "by", "with", "from",
		"into", "onto", "over", "under", "about", "as", "is", "are", "was",
		"were", "be", "been", "being", "this", "that", "these", "those",
		"it", "its", "you", "your", "yours", "i", "we", "our", "ours",
		"they", "their", "he", "she", "his", "her", "them", "not", "no",
		"do", "does", "did", "can", "could", "should", "would", "will",
		"shall", "may", "might", "must", "also", "than", "such", "any",
		"all", "some", "each", "every", "other", "another", "more", "most",
		"much", "many", "only", "just", "very", "so", "too", "own", "same",
		"here", "there", "what", "which", "who", "whom", "how", "why",
		"use", "uses", "using", "used", "skill", "skills", "tool", "tools",
		"user", "users", "claude", "agent", "agents", "code", "file",
		"files", "task", "tasks", "want", "wants", "need", "needs",
	}
	m := make(map[string]struct{}, len(words))
	for _, w := range words {
		m[w] = struct{}{}
	}
	return m
}

// Finding is one under- or over-triggered skill report entry.
type Finding struct {
	SkillName string
	Count     int
	Evidence  string
}

// Report holds the fidelity scan results.
type Report struct {
	UnderTriggered []Finding
	OverTriggered  []Finding
}

// skillKeywords is the TF-IDF-weighted keyword profile for one skill.
type skillKeywords struct {
	name    string             // lowercased
	weights map[string]float64 // token (len >= 4, not a stopword) -> IDF weight
	phrases []string           // lowercased multi-word phrases from "Triggers on:" sections
}

func extractTriggerPhrases(description string) []string {
	loc := triggerSectionRe.FindStringIndex(description)
	if loc == nil {
		return nil
	}
	tail := description[loc[1]:]
	seen := map[string]struct{}{}
	var phrases []string
	for _, m := range quotedPhraseRe.FindAllStringSubmatch(tail, -1) {
		p := strings.ToLower(strings.TrimSpace(m[1]))
		if p == "" {
			continue
		}
		if _, ok := seen[p]; !ok {
			seen[p] = struct{}{}
			phrases = append(phrases, p)
		}
	}
	return phrases
}

// skillTokens extracts the len>=4, non-stopword token set from a
// skill's "name description" text.
func skillTokens(skill models.SkillDefinition) map[string]struct{} {
	text := strings.ToLower(skill.Name + " " + skill.Description)
	tokens := wordRe.FindAllString(text, -1)
	set := make(map[string]struct{})
	for _, t := range tokens {
		if len(t) < 4 {
			continue
		}
		if _, stop := stopwords[t]; stop {
			continue
		}
		set[t] = struct{}{}
	}
	return set
}

// buildTFIDFKeywords computes IDF weights for every skill's keyword
// set. IDF is computed over the corpus of skill descriptions:
// idf(t) = ln(N / df(t)) + 1, where N is the number of skills and
// df(t) is the number of skill descriptions containing token t at
// least once. The "+1" floor keeps universally-shared tokens (df==N)
// from scoring zero — a skill still needs 2+ matched tokens under the
// default threshold, so this only affects borderline single-token
// cases.
func buildTFIDFKeywords(skills []models.SkillDefinition) map[string]skillKeywords {
	n := float64(len(skills))
	tokenSets := make([]map[string]struct{}, len(skills))
	df := map[string]int{}
	for i, s := range skills {
		tokenSets[i] = skillTokens(s)
		for t := range tokenSets[i] {
			df[t]++
		}
	}

	idf := map[string]float64{}
	for t, count := range df {
		idf[t] = math.Log(n/float64(count)) + 1
	}

	result := make(map[string]skillKeywords, len(skills))
	for i, s := range skills {
		weights := make(map[string]float64, len(tokenSets[i]))
		for t := range tokenSets[i] {
			weights[t] = idf[t]
		}
		result[s.Name] = skillKeywords{
			name:    strings.ToLower(s.Name),
			weights: weights,
			phrases: extractTriggerPhrases(s.Description),
		}
	}
	return result
}

func nameMatches(textLower, nameLower string) bool {
	if nameLower == "" {
		return false
	}
	normalize := func(s string) string {
		return regexp.MustCompile(`[-_]+`).ReplaceAllString(s, " ")
	}
	normalizedName := strings.TrimSpace(normalize(nameLower))
	if normalizedName == "" {
		return false
	}
	normalizedText := normalize(textLower)
	return strings.Contains(" "+normalizedText+" ", " "+normalizedName+" ")
}

// tfidfThreshold is the minimum summed IDF weight of matched tokens
// for a prompt to "plausibly match" a skill via keyword overlap alone
// (name match and quoted trigger phrases always short-circuit to a
// match regardless of this threshold). A single rare token (high IDF)
// can otherwise clear a weight-only threshold on its own — e.g. one
// distinctive but generic word landing in an unrelated prompt — so a
// minimum distinct-token-count floor (minMatchedTokens) is required in
// addition to the weight sum.
//
// Tuned empirically against the live ~/.claude/projects corpus (357
// discovered skills, via internal/fidelity/sweep_test.go). Weight-only
// scoring at any threshold left the under-triggered list flat around
// 340-357 (single-token matches dominating, no better than Python's
// raw ">=2 keyword hits" rule which flagged 346/357). Requiring >=6
// distinct matched tokens plus a weight sum of >=25.0 brings the
// keyword-overlap contribution down to near zero, landing the
// under-triggered list at its practical floor of ~53 skills on this
// corpus.
//
// That floor is *not* further reducible by tuning these two
// constants alone: it is dominated by the spec-mandated,
// verbatim-preserved instant name-match rule (nameMatches below)
// colliding with several skill names that are also common English
// words in this corpus's domain — "security", "research", "terraform",
// "okta", "teleport", "gh", "report", "todo", "vpn", "gcp", "mfa" —
// each of which legitimately appears in a large fraction of this PSEC
// engineer's prompts regardless of skill relevance. In isolation
// (keyword scoring fully disabled) the name-match rule alone produces
// 51 under-triggered skills at the old fixed count>=3 cutoff. This is
// a real corpus characteristic of a single, topically-concentrated
// user, not a tuning defect.
//
// Lever that actually closes the gap: minSessionCount (below). A
// common-word skill name (e.g. "security") legitimately matches a
// prompt-only-once-or-twice pattern across many sessions, but only
// fires as a *sustained* signal — many distinct matching sessions —
// for skills the engineer genuinely reaches for repeatedly; one-off
// common-word collisions rarely clear a higher session-count bar.
// Raising the report's minimum-session-count cutoff from a fixed 3 to
// a tunable, env-overridable default of 8 (mirroring the sibling Rust
// implementation's fix for the identical structural floor, including
// its env var names for cross-implementation consistency) brings the
// under-triggered list from ~53 down to 26 on this corpus — within the
// spec's 10-30 target band — without weakening the name-match rule
// itself. Override via SKILLSCOPE_MIN_SESSION_COUNT /
// SKILLSCOPE_TFIDF_THRESHOLD if a different corpus lands outside that
// band.
var (
	tfidfThreshold   = 25.0
	minMatchedTokens = 6

	// minSessionCount is the minimum number of distinct matching
	// sessions a skill needs to appear in the under-triggered report.
	// Resolved from SKILLSCOPE_MIN_SESSION_COUNT at the top of Run
	// (not here, at package-init time) so it composes with
	// t.Setenv-based testing.
	minSessionCount = defaultMinSessionCount
)

const (
	defaultMinSessionCount = 8

	// envMinSessionCount and envTFIDFThreshold match the names used by
	// the sibling Rust implementation (poc/rust-ratatui), so a user
	// tuning one implementation's fidelity thresholds gets the same
	// effect setting the same env var against either binary.
	envMinSessionCount = "SKILLSCOPE_MIN_SESSION_COUNT"
	envTFIDFThreshold  = "SKILLSCOPE_TFIDF_THRESHOLD"
)

// envInt reads an environment variable as an int, falling back to def
// if the variable is unset or unparseable.
func envInt(name string, def int) int {
	raw := os.Getenv(name)
	if raw == "" {
		return def
	}
	v, err := strconv.Atoi(raw)
	if err != nil {
		return def
	}
	return v
}

// envFloat reads an environment variable as a float64, falling back to
// def if the variable is unset or unparseable.
func envFloat(name string, def float64) float64 {
	raw := os.Getenv(name)
	if raw == "" {
		return def
	}
	v, err := strconv.ParseFloat(raw, 64)
	if err != nil {
		return def
	}
	return v
}

// promptMatchesSkill reports whether a prompt "plausibly matches" a
// skill: contains the skill name as a word, OR contains a quoted
// trigger phrase from a "Triggers on:" section, OR the prompt overlaps
// at least minMatchedTokens distinct keywords from the skill's profile
// with a summed IDF weight meeting tfidfThreshold.
func promptMatchesSkill(textLower string, tokens map[string]struct{}, kw skillKeywords) bool {
	if nameMatches(textLower, kw.name) {
		return true
	}
	for _, phrase := range kw.phrases {
		if strings.Contains(textLower, phrase) {
			return true
		}
	}
	var score float64
	var matched int
	for t := range tokens {
		if w, ok := kw.weights[t]; ok {
			score += w
			matched++
		}
	}
	return matched >= minMatchedTokens && score >= tfidfThreshold
}

// isNoisePrompt applies Feature 1's additional prompt filters, beyond
// the core parser's transcript-level filters: plan-mode injections,
// teammate-message wrapper text, and prompts too short to carry real
// intent.
func isNoisePrompt(text string) bool {
	if len(text) < 15 {
		return true
	}
	if strings.HasPrefix(text, "Implement the following plan:") {
		return true
	}
	if strings.Contains(text, "<teammate-message") {
		return true
	}
	return false
}

// Options configures RunFidelity.
type Options struct {
	ProjectsDir           string
	SkillsDirs            []string
	PluginMarketplacesDir string
	Since                 *timeFilter
}

// timeFilter is a minimal since-cutoff; kept as its own type so callers
// don't need to import time in the common case of an unfiltered run.
type timeFilter struct {
	unixNano int64
}

// NewSinceFilter builds a since-cutoff from a Unix-nanosecond timestamp.
func NewSinceFilter(unixNano int64) *timeFilter { return &timeFilter{unixNano: unixNano} }

// Run executes the fidelity scan: discover skills, stream invocations
// and prompts, correlate per-session, and bucket into under/over
// triggered findings.
func Run(opts Options) (Report, error) {
	// Resolved at call time (not package init) so tests can use
	// t.Setenv to exercise both the override and fallback paths.
	tfidfThreshold = envFloat(envTFIDFThreshold, 25.0)
	minSessionCount = envInt(envMinSessionCount, defaultMinSessionCount)

	skills := DiscoverSkills(opts.SkillsDirs, opts.PluginMarketplacesDir)
	skillKW := buildTFIDFKeywords(skills)

	sessionInvocations := map[string]map[string]struct{}{}
	err := parser.IterInvocations(opts.ProjectsDir, func(inv models.SkillInvocation) bool {
		if opts.Since != nil && inv.Timestamp.UnixNano() < opts.Since.unixNano {
			return true
		}
		set, ok := sessionInvocations[inv.SessionID]
		if !ok {
			set = map[string]struct{}{}
			sessionInvocations[inv.SessionID] = set
		}
		set[inv.SkillName] = struct{}{}
		return true
	})
	if err != nil {
		return Report{}, err
	}

	sessionPrompts := map[string][]models.UserPrompt{}
	err = parser.IterUserPrompts(opts.ProjectsDir, func(p models.UserPrompt) bool {
		if opts.Since != nil && p.Timestamp.UnixNano() < opts.Since.unixNano {
			return true
		}
		if isNoisePrompt(p.Text) {
			return true
		}
		sessionPrompts[p.SessionID] = append(sessionPrompts[p.SessionID], p)
		return true
	})
	if err != nil {
		return Report{}, err
	}

	underCounts := map[string]int{}
	underExamples := map[string][]string{}
	overCounts := map[string]int{}
	overExamples := map[string][]string{}

	allSessions := map[string]struct{}{}
	for sid := range sessionInvocations {
		allSessions[sid] = struct{}{}
	}
	for sid := range sessionPrompts {
		allSessions[sid] = struct{}{}
	}

	for sessionID := range allSessions {
		prompts := sessionPrompts[sessionID]
		invoked := sessionInvocations[sessionID]

		sessionMatched := map[string]struct{}{}
		matchedSnippet := map[string]string{}
		for _, prompt := range prompts {
			textLower := strings.ToLower(prompt.Text)
			tokenList := wordRe.FindAllString(textLower, -1)
			tokens := make(map[string]struct{}, len(tokenList))
			for _, t := range tokenList {
				tokens[t] = struct{}{}
			}
			for skillName, kw := range skillKW {
				if _, already := sessionMatched[skillName]; already {
					continue
				}
				if promptMatchesSkill(textLower, tokens, kw) {
					sessionMatched[skillName] = struct{}{}
					snippet := prompt.Text
					if len(snippet) > 120 {
						snippet = snippet[:120]
					}
					matchedSnippet[skillName] = snippet
				}
			}
		}

		for skillName := range sessionMatched {
			if _, ok := invoked[skillName]; ok {
				continue
			}
			underCounts[skillName]++
			if len(underExamples[skillName]) < 3 {
				underExamples[skillName] = append(underExamples[skillName], matchedSnippet[skillName])
			}
		}

		for skillName := range invoked {
			if _, ok := sessionMatched[skillName]; ok {
				continue
			}
			overCounts[skillName]++
			snippet := "(no user prompt text in session)"
			if len(prompts) > 0 {
				snippet = prompts[0].Text
				if len(snippet) > 120 {
					snippet = snippet[:120]
				}
			}
			if len(overExamples[skillName]) < 3 {
				overExamples[skillName] = append(overExamples[skillName], snippet)
			}
		}
	}

	var under []Finding
	for name, count := range underCounts {
		if count < minSessionCount {
			continue
		}
		under = append(under, Finding{SkillName: name, Count: count, Evidence: strings.Join(underExamples[name], " | ")})
	}
	sort.SliceStable(under, func(i, j int) bool { return under[i].Count > under[j].Count })

	var over []Finding
	for name, count := range overCounts {
		over = append(over, Finding{SkillName: name, Count: count, Evidence: strings.Join(overExamples[name], " | ")})
	}
	sort.SliceStable(over, func(i, j int) bool { return over[i].Count > over[j].Count })

	return Report{UnderTriggered: under, OverTriggered: over}, nil
}

File: experiments/python/src/skillscope/fidelity.py
"""Trigger-fidelity eval layer: skill discovery + offline heuristic classification.

No LLM calls. Correlates real user prompts against skill frontmatter descriptions
to flag two failure modes:

- under-triggering: a skill's description plausibly matches a prompt, but the
  skill was never invoked in that session
- over-triggering: a skill was invoked in a session, but no prompt in that
  session plausibly matches its description
"""

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Any

import yaml

from skillscope.models import SkillDefinition, UserPrompt

DEFAULT_SKILLS_DIRS: list[Path] = [
    Path.home() / ".agents" / "skills",
    Path.home() / ".claude" / "skills",
]
PLUGIN_MARKETPLACES_DIR = Path.home() / ".claude" / "plugins" / "marketplaces"

_WORD_RE = re.compile(r"[a-z0-9][a-z0-9_-]*")
_TRIGGER_SECTION_RE = re.compile(r"triggers?\s+on\s*:", re.IGNORECASE)
_QUOTED_PHRASE_RE = re.compile(r'"([^"]{2,60})"')

_STOPWORDS: frozenset[str] = frozenset(
    {
        # common English function words
        "a", "an", "the", "and", "or", "but", "if", "then", "else", "when",
        "while", "for", "to", "of", "in", "on", "at", "by", "with", "from",
        "into", "onto", "over", "under", "about", "as", "is", "are", "was",
        "were", "be", "been", "being", "this", "that", "these", "those",
        "it", "its", "you", "your", "yours", "i", "we", "our", "ours",
        "they", "their", "he", "she", "his", "her", "them", "not", "no",
        "do", "does", "did", "can", "could", "should", "would", "will",
        "shall", "may", "might", "must", "also", "than", "such", "any",
        "all", "some", "each", "every", "other", "another", "more", "most",
        "much", "many", "only", "just", "very", "so", "too", "own", "same",
        "here", "there", "what", "which", "who", "whom", "how", "why",
        # generic / domain-neutral words called out in the spec
        "use", "uses", "using", "used", "skill", "skills", "tool", "tools",
        "user", "users", "claude", "agent", "agents", "code", "file",
        "files", "task", "tasks", "will", "want", "wants", "need", "needs",
    }
)


@dataclass
class FidelityFinding:
    skill_name: str
    count: int
    evidence: str


@dataclass
class FidelityReport:
    under_triggered: list[FidelityFinding]
    over_triggered: list[FidelityFinding]


@dataclass(frozen=True)
class _SkillKeywords:
    name: str  # lowercased
    words: frozenset[str]  # single tokens, len >= 4, stopwords removed
    phrases: frozenset[str]  # lowercased multi-word phrases from "Triggers on:" sections


def _parse_frontmatter(content: str) -> dict[str, Any] | None:
    if not content.startswith("---"):
        return None
    parts = content.split("---", 2)
    if len(parts) < 3:
        return None
    try:
        data = yaml.safe_load(parts[1])
    except yaml.YAMLError:
        return None
    if not isinstance(data, dict):
        return None
    return data


def _iter_skill_md_candidates(
    skills_dirs: list[Path],
) -> list[tuple[Path, str]]:
    """Yield (SKILL.md path, source) pairs, deduped by resolved path."""
    seen: set[Path] = set()
    candidates: list[tuple[Path, str]] = []

    for skills_dir in skills_dirs:
        if not skills_dir.is_dir():
            continue
        for entry in sorted(skills_dir.iterdir()):
            skill_md = entry / "SKILL.md"
            if not skill_md.is_file():
                continue
            resolved = skill_md.resolve()
            if resolved in seen:
                continue
            seen.add(resolved)
            candidates.append((skill_md, "user"))

    if PLUGIN_MARKETPLACES_DIR.is_dir():
        for skill_md in sorted(PLUGIN_MARKETPLACES_DIR.glob("*/*/skills/*/SKILL.md")):
            resolved = skill_md.resolve()
            if resolved in seen:
                continue
            seen.add(resolved)
            candidates.append((skill_md, "plugin"))

    return candidates


def discover_skills(skills_dirs: list[Path] | None = None) -> list[SkillDefinition]:
    """Scan skill directories for SKILL.md files and parse their frontmatter.

    Files without valid YAML frontmatter are skipped gracefully.
    """
    dirs = skills_dirs if skills_dirs is not None else DEFAULT_SKILLS_DIRS
    results: list[SkillDefinition] = []

    for skill_md, source in _iter_skill_md_candidates(dirs):
        try:
            content = skill_md.read_text(encoding="utf-8")
        except OSError:
            continue
        frontmatter = _parse_frontmatter(content)
        if frontmatter is None:
            continue
        name = str(frontmatter.get("name") or skill_md.parent.name)
        description = str(frontmatter.get("description") or "")
        results.append(
            SkillDefinition(
                name=name,
                description=description,
                path=str(skill_md.parent),
                source=source,
            )
        )

    return results


def _extract_trigger_phrases(description: str) -> frozenset[str]:
    match = _TRIGGER_SECTION_RE.search(description)
    if not match:
        return frozenset()
    tail = description[match.end():]
    phrases = {p.strip().lower() for p in _QUOTED_PHRASE_RE.findall(tail) if p.strip()}
    return frozenset(phrases)


def _build_skill_keywords(skill: SkillDefinition) -> _SkillKeywords:
    phrases = _extract_trigger_phrases(skill.description)
    text = f"{skill.name} {skill.description}".lower()
    tokens = _WORD_RE.findall(text)
    words = frozenset(t for t in tokens if len(t) >= 4 and t not in _STOPWORDS)
    return _SkillKeywords(name=skill.name.lower(), words=words, phrases=phrases)


def _name_matches(text_lower: str, name_lower: str) -> bool:
    if not name_lower:
        return False
    normalized_name = re.sub(r"[-_]+", " ", name_lower).strip()
    if not normalized_name:
        return False
    normalized_text = re.sub(r"[-_]+", " ", text_lower)
    return f" {normalized_name} " in f" {normalized_text} "


def prompt_matches_skill(
    text_lower: str, tokens: frozenset[str], keywords: _SkillKeywords
) -> bool:
    """A prompt "plausibly matches" a skill per the heuristic rules:

    contains the skill name as a word, OR contains >= 2 distinct keywords
    (len >= 4) from the skill's keyword set, OR contains a quoted trigger
    phrase declared in a "Triggers on:" section of the description.
    """
    if _name_matches(text_lower, keywords.name):
        return True
    if any(phrase in text_lower for phrase in keywords.phrases):
        return True
    return len(tokens & keywords.words) >= 2


def run_fidelity(
    projects_dir: Path | None = None,
    skills_dirs: list[Path] | None = None,
    since: datetime | None = None,
) -> FidelityReport:
    from skillscope.parser import iter_invocations, iter_user_prompts

    skills = discover_skills(skills_dirs)
    skill_keywords = {s.name: _build_skill_keywords(s) for s in skills}

    # iter_invocations/iter_user_prompts default projects_dir themselves;
    # only override when the caller passed an explicit path.
    parser_kwargs: dict[str, Path] = {}
    if projects_dir is not None:
        parser_kwargs["projects_dir"] = projects_dir

    session_invocations: dict[str, set[str]] = defaultdict(set)
    for inv in iter_invocations(**parser_kwargs):
        if since is not None and inv.timestamp < since:
            continue
        session_invocations[inv.session_id].add(inv.skill_name)

    session_prompts: dict[str, list[UserPrompt]] = defaultdict(list)
    for prompt in iter_user_prompts(**parser_kwargs):
        if since is not None and prompt.timestamp < since:
            continue
        session_prompts[prompt.session_id].append(prompt)

    under_counts: dict[str, int] = defaultdict(int)
    under_examples: dict[str, list[str]] = defaultdict(list)
    over_counts: dict[str, int] = defaultdict(int)
    over_examples: dict[str, list[str]] = defaultdict(list)

    all_sessions = set(session_invocations) | set(session_prompts)
    for session_id in all_sessions:
        prompts = session_prompts.get(session_id, [])
        invoked = session_invocations.get(session_id, set())

        session_matched: set[str] = set()
        matched_snippet: dict[str, str] = {}
        for prompt in prompts:
            text_lower = prompt.text.lower()
            tokens = frozenset(_WORD_RE.findall(text_lower))
            for skill_name, keywords in skill_keywords.items():
                if skill_name in session_matched:
                    continue
                if prompt_matches_skill(text_lower, tokens, keywords):
                    session_matched.add(skill_name)
                    matched_snippet[skill_name] = prompt.text[:120]

        for skill_name in session_matched - invoked:
            under_counts[skill_name] += 1
            if len(under_examples[skill_name]) < 3:
                under_examples[skill_name].append(matched_snippet[skill_name])

        for skill_name in invoked - session_matched:
            over_counts[skill_name] += 1
            snippet = prompts[0].text[:120] if prompts else "(no user prompt text in session)"
            if len(over_examples[skill_name]) < 3:
                over_examples[skill_name].append(snippet)

    under_triggered = [
        FidelityFinding(skill_name=name, count=count, evidence=" | ".join(under_examples[name]))
        for name, count in under_counts.items()
        if count >= 3
    ]
    under_triggered.sort(key=lambda f: f.count, reverse=True)

    over_triggered = [
        FidelityFinding(skill_name=name, count=count, evidence=" | ".join(over_examples[name]))
        for name, count in over_counts.items()
    ]
    over_triggered.sort(key=lambda f: f.count, reverse=True)

    return FidelityReport(under_triggered=under_triggered, over_triggered=over_triggered)
